FEA Add DecisionBoundaryDisplay

thomasjpfan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->
This PR adds `BoundaryDecisionDisplay` to the inspection module.

#### What does this implement/fix? Explain your changes.
Many examples were touched because they can take advantage of the new function.


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

thomasjpfan committed:
WIP

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
ENH Completely adds decision boundary

thomasjpfan committed:
TST Adds more tests

thomasjpfan committed:
STY Linting

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
BUG Fix

jnothman commented:
I appreciate the benefit to reduce code duplication in our examples, but is
this useful outside of teaching?

thomasjpfan commented:
I had the same discussion with @amueller a few months ago regarding including this plot function. I think it is useful only in teaching, but we do a bunch of teaching in our examples.

amueller commented:
and it will make life much easier for those teaching with sklearn (though that's admittedly a slippery slope). It certainly declutters the examples.

amueller commented:
supersedes #5070

NicolasHug commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,193 @@
+    Parameters
+    ----------
+    xx0 : ndarray of shape (grid_resolution, grid_resolution)
+        First output of meshgrid.
```
Maybe we can also crosslink to numpy docs
```suggestion
        First output of np.meshgrid.
```

NicolasHug commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,193 @@
+                           plot_method='contourf',
+                           response_method='auto',
+                           ax=None, **kwargs):
+    """Plot Decision Boundary.
```
This would need a UG link or at least something like "please see examples below for usage"

NicolasHug commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,193 @@
+from ...utils import check_matplotlib_support
+
+
+def _check_boundary_response_method(estimator, response_method):
```
Haven't checked in details but don't we have this logic duplicated in various places, modulo some minor differences?

NicolasHug commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,193 @@
+        Parameters
+        ----------
+        plot_method : {'contourf', 'contour', 'pcolormesh'}, default='contourf'
+            Plotting method to call when plotting the response.
```
maybe add "please refer to matplotlib documentation" or something like that

NicolasHug commented on sklearn/inspection/_plot/tests/test_plot_decision_boundary.py:
```diff
@@ -0,0 +1,122 @@
+    X = X[:, [0, 1]]
+    lr = LogisticRegression().fit(X, y)
+
+    msg = ("multiclass classifers are only supported when "
```
```suggestion
    msg = ("multiclass classifiers are only supported when "
```

NicolasHug reviewed (commented):
+1 for this.

Not just useful for teaching but also for people learning / experimenting on their own, which represent a huge part of our users I believe

A few comments but looks good.

@thomasjpfan There seem however to be a significant difference between the plots in the examples eg

 https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html
versus
https://88180-843222-gh.circle-artifacts.com/0/doc/auto_examples/classification/plot_classifier_comparison.html

amueller commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,193 @@
+            raise ValueError(error_msg.format(response_method,
+                                              estimator.__class__.__name__))
+        return getattr(estimator, response_method)
+    elif hasattr(estimator, 'predict_proba'):
```
right now "classifier comparison" does decision_function, then predict_proba. For plotting I think that actually makes sense since a softmax will always compress down the boundary.
I assume you got this from metrics, where having probabilities might be preferred.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
CLN Update response_method order

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
CLN Adds links to external libraries

thomasjpfan committed:
CLN Adds reference to quad*

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,205 @@
+        return self
+
+
+def plot_decision_boundary(estimator, X, grid_resolution=100, eps=1.0,
```
I am a bit uncomfortable to add this function in the inspection module because I don't think that anyone should use this for inspecting or draw any conclusion on a subset of features. However, I see the benefit of having this function as a miscellaneous function for making example less complex and for teaching.

thomasjpfan replied:
To clarify, are you suggesting the `sklearn.utils` module? I am +0.5 on that.

The counter argument would be "Since we are using it to teach, there must be some conclusions we can draw from it...otherwise we wont use it to teach"

I kind of loosely defined inspection as "things you can do after your model has been fitted".

glemaitre replied:
> To clarify, are you suggesting the sklearn.utils module? I am +0.5 on that.

I don't know what would be the best module to accommodate it. `utils` seems the most suitable with the current tree-view for me. Otherwise, it would be in a new module.

>The counter argument would be "Since we are using it to teach, there must be some conclusions we can draw from it...otherwise we wont use it to teach" I kind of loosely defined inspection as "things you can do after your model has been fitted".

I personally used it to give intuitions about how an algorithm works in low dimensional space. For instance, it makes it easy to understand that our tree finds the best split along a specific feature (perpendicular split to an axis) or what is the meaning of `intercept_` or `coef_` in linear models.

I do not consider those as inspection/interpretation (because you should not use it in practice :)). Indeed, I am scared of the following: by making it easy, one can select a pair of features and make some kind of interpretation regarding what a model learnt.

On the opposite, I would not mind having the `plot_tree` in inspection because it depicts exactly what your model learnt.

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,205 @@
+    estimator : estimator instance
+        Trained estimator.
+
+    X : array-like of shape (n_samples, 2)
```
Since we are passing `X`, maybe we could have an option to make the `scatter` plot at the same time and to set the x_min/x_max/y_min/y_max limits. Basically, I think that most of the time, we plotting both the boundary and the scatter plot.

thomasjpfan replied:
I included this at the start, but I felt like it inflated the API. As in, we may need a `scater_kwargs` in the `plot_decisions_boundary`.

In its current form, plotting the scatter plot isn't too bad:

```py
fig, ax = plt.subplots()
plot_decision_boundary(est, X, y, ax=ax)
ax.scatter(X[:, 0], X[:, 1], ...)
```

It's common to want to adjust the alpha when there are too many points (or subsample).

glemaitre replied:
I think that it would be fine to have `scatter_kwargs` in fact, I would like to have to write the minimum lines of code, especially `xlim`, `ylim`, `xlabel`, `ylabel`, `alpha`, etc., for the largest number of our examples. Then, when one wants to fine-tune it will come back as efficient to write some matplotlib.

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,205 @@
+        Axes object to plot on. If `None`, a new figure and axes is
+        created.
+
+    **kwargs : dict
```
I am also thinking that we should have a way to decorate xlabel and ylabel if we are passing a dataframe using the column names.

glemaitre commented on sklearn/inspection/_plot/tests/test_plot_decision_boundary.py:
```diff
@@ -0,0 +1,158 @@
+@pytest.fixture(scope="module")
+def data():
+    X, y = make_classification(n_informative=1, n_redundant=1,
+                               n_clusters_per_class=1, n_features=2)
```
```suggestion
                               n_clusters_per_class=1, n_features=2
                               random_state=42)
```

glemaitre commented on sklearn/setup.py:
```diff
@@ -39,6 +39,8 @@ def configuration(parent_package='', top_path=None):
     config.add_subpackage('impute/tests')
     config.add_subpackage('inspection')
     config.add_subpackage('inspection/tests')
+    config.add_subpackage('inspection/_plot')
```
I don't think this is necessary because I added a `setup.py` in the `inspection` module.

glemaitre reviewed (commented):
Codewise, I think this completely fine. Just some questions regarding the API, the location of the function, and what we expect when we are using it.

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
BUG Fix

amueller commented:
CI failing

thomasjpfan commented:
CI failure is unrelated (But should be fixed with a merge in this case)

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

amueller commented:
I'm fine putting it in ``inspection`` or ``utils``. I think I agree with @glemaitre that ``utils`` might be a bit more natural.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
CLN Move to utils

thomasjpfan committed:
BUG Fix

thomasjpfan committed:
ENH Move to utils

thomasjpfan commented:
Moved to `sklearn.utils.plot` to avoid circular dependencies with `sklearn.utils.__init__.py`.

If we are not okay with this, I will move `check_matplotlib_support` and `_safe_indexing` into its own files and so we can have `sklearn.utils.plot_decision_boundary`.

thomasjpfan committed:
BLD Fixes build error

thomasjpfan committed:
FIX Bug

amueller commented:
still failing ;)

NicolasHug commented:
Sorry to be late but I feel like utils isn't the right place for this. If you look at the `utils` API ref, all we have there are developer tools and `plot_decision_boundary` doesn't seem to fit there.

Also in the original discussion https://github.com/scikit-learn/scikit-learn/issues/13448, we were concerned about scattering the plotting tools in too many different modules.

I would agree with @thomasjpfan that conclusions can be drawn from a decision boundary, and that's fine. (Arguably, you can never draw any conclusion about a PDP anyway because your features are never independent).

I understand `sklearn.inspection` as being a broad term for "looking at" what a model does. `plot` was actually a suggested alternative for the name. So IMO this is the right place for `plot_decision_boundary`.

jnothman commented:
+1 for inspection.
In terms of use beyond teaching... do people plot decision boundaries on 2
dimensional representations (e.g. PCs) of higher-order decision functions?

thomasjpfan committed:
CLN Move back to inspection

jnothman commented:
I'm not planning to review this in detail, but it does look ready to go, assuming there is consensus that we want such teaching tools in the library.

glemaitre commented:
@thomasjpfan Let's move forward with this visualization. I was the one blocking here. Could you solve the conflict and add an entry in what's new.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

thomasjpfan committed:
DOC Adds whats new

glemaitre commented:
ping @NicolasHug if you want to have another look and go towards merging?

glemaitre committed:
merge master

glemaitre commented:
I solve the conflicts. @NicolasHug do you mind to give a fresh look.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into plot_decision_bound

cmarmo commented:
@NicolasHug do you mind having a look here? Thanks!

NicolasHug commented:
Sorry I won't have much time for big reviews like these in the coming days

glemaitre commented:
So with the new plotting API, we should only have the Display and in this case, `.from_estimator` only should be implemented.

glemaitre commented:
I am moving this feature to the next milestone. This is a nice to have feature but we might want to focus on the other priorities before 0.24. If time allows, we can always come back on it.

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/thomasjpfan/16061

glemaitre committed:
API move to new plotting API

glemaitre committed:
iter

glemaitre committed:
rename file and avoid warning

glemaitre committed:
FEA allow to set x-/y-label

glemaitre committed:
DOC fix

glemaitre committed:
avoid to set axis outside display

glemaitre commented:
@thomasjpfan I am currently working on adapting the code to the new plotting API. It is almost there. I need to look at the documentation mainly but the rest should be fine.

For this helper function, I think that `from_predictions` does not make that much sense since you need to construct the meshgrid beforehand.

glemaitre committed:
DOC add example and see also

glemaitre reviewed (approved):
@thomasjpfan I think this is good to have a new look.
It looks good on my side.

glemaitre commented on .pre-commit-config.yaml:
```diff
@@ -9,6 +9,7 @@ repos:
     rev: 21.6b0
     hooks:
     -   id: black
+        exclude: ^examples
```
Otherwise, pre-commit is blackifying the examples as well. We might want to avoid this for the moment and minimize the diff.

glemaitre committed:
iter

adrinjalali commented:
@thomasjpfan or @amueller  wanna have another look here? It's been around for a while, and seems like you want it in the next release.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
REV Revert examples for now

thomasjpfan commented:
Thank you @glemaitre for migrating this to the `from_estimator` API!

I do not think this is a blocker for 1.0 release. (It's really mostly a teaching tool to get a "feel" for an estimator)

ogrisel commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,318 @@
+
+        if response.ndim != 1:
+            if response.shape[1] != 2:
+                raise ValueError(
+                    "Multiclass classifiers are only supported when "
+                    "response_method='predict'"
+                )
```
Also I am not sure what will happen with a multi-label classifier or a multi-output multiclass classifier or a multioutput regressor.

ogrisel commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,318 @@
+                    "response_method='predict'"
+                )
+            response = response[:, 1]
+
```
I think that if `response.ndim == 1` and `pred_func.__name__ == "predict"`and `hasattr(estimator, "classes_")` we should use the latter to re-encode `response` using `classes_` to get integers starting at 0. and use the class labels in the plot with a legend.

ogrisel committed:
Merge branch 'main' into plot_decision_bound

ogrisel reviewed (approved):
This is already a net improvement so I am ok for merging this as it is. However there are plenty of things that could be improved, possibly in follow-up PRs.

I agree this would be really useful for teaching! I hadn't realized we had this `meshgrid/contourf` code pattern repeated so many times in our examples.

ogrisel commented:
BTW I check the rendering of the updated example and they look mostly unchanged.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
ENH Better validation errors

thomasjpfan commented:
I am starting to wonder if decision boundary should be included. It is not too different from two-way partial dependence  that only works on 2 features. For our examples that needs the decision boundaries, we can add a kwargs to partial dependence to remove the contour lines.

![Screen Shot 2021-09-06 at 12 37 22 PM](https://user-images.githubusercontent.com/5402633/132245537-740124ab-36e6-4f8b-9abd-e21c00bbbbea.png)

WDYT @glemaitre @ogrisel ?

ogrisel commented:
That's a good point but I have the feeling that it's still useful to have a dedicated, educational function for the 2D case.

- it allows us to support multiclass decision boundaries for the `predict` method (hard classification) which partial decision points;
- it makes it possible choose colormaps that have white at the class seperation boundary (e.g. 0 for `decision_function` or 0.5 for `predict_proba`). That would be too cumbersome for change partial decision plots to do this;
- understanding what partial decision plots is a cognitive overhead for most examples where we just want to plot the true decision boundaries of a model with exactly only 2 input features.

ogrisel commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,324 @@
+            )
+
+        possible_plot_methods = ("contourf", "contour", "pcolormesh")
+        if plot_method not in possible_plot_methods:
+            avaliable_methods = ", ".join(possible_plot_methods)
+            raise ValueError(
+                f"plot_method must be one of {avaliable_methods}. "
+                f"Got {plot_method} instead."
+            )
```
typo:

```suggestion
        if plot_method not in possible_plot_methods:
            available_methods = ", ".join(possible_plot_methods)
            raise ValueError(
                f"plot_method must be one of {available_methods}. "
                f"Got {plot_method} instead."
            )
```

glemaitre commented:
There some glitches in the documentation. I added some other comments that are valid if we want to follow the same pattern than the other `plot_*` functions

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -0,0 +1,324 @@
+        )
+
+        pred_func = _check_boundary_response_method(estimator, response_method)
+        response = pred_func(np.c_[xx0.ravel(), xx1.ravel()])
```
We will have an issue here in case that `response_method="predict"` and that the we are in classification with non-encoded labels. The response can be non-numeric values that cannot be converted into numeric values. The workaround would be to initialized a `LabelEncoder` with the `classes_` fitted attributes that are `estimator.classes_`. I think that it would be enough

glemaitre replied:
There is still this issue. If I remember I triggered the bug while using the class.

thomasjpfan replied:
Fixed in [`b864194` (#16061)](https://github.com/scikit-learn/scikit-learn/pull/16061/commits/b864194b551173689aa6ad07341da01a4f6b54f7). It uses `LabelEncoder` as you suggested. `LabelEncoder` is a bit more robust just in case the classes were integer but did not start with 0. i.e. `classes_ = [1, 3]`.

There is already a test for the string use case here: https://github.com/thomasjpfan/scikit-learn/blob/b864194b551173689aa6ad07341da01a4f6b54f7/sklearn/inspection/_plot/tests/test_boundary_decision_display.py#L306

thomasjpfan committed:
ENH Remvoe

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
ENH Update avaliable methods

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
DOC Move to 1.1

thomasjpfan committed:
WIP

thomasjpfan committed:
FIX Support string as targets

thomasjpfan committed:
STY Fix black formatting

lesteve committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into pr/16061

lesteve commented:
There were 2 approvals, I fixed the conflicts and pushed, let's see what happens

lesteve committed:
Fix whats_new

thomasjpfan committed:
DOC Fixes whats nwe

thomasjpfan committed:
ENH Improve auto behavior for multiclass problems

thomasjpfan committed:
FIX Removes unneeded code

thomasjpfan commented:
> There were 2 approvals, I fixed the conflicts and pushed, let's see what happens

I'm giving this PR another pass. I see there are still some comments left to address.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
FIX Adds classes to InductiveClusterer

thomasjpfan committed:
ENH Do not require classes

thomasjpfan committed:
DOC Adds TODO

thomasjpfan committed:
ENH Improve error message for unsupported estimators

thomasjpfan committed:
CLN Remove comment

thomasjpfan commented:
I updated this PR addressing most of the remaining concerns. The remaining comment is to support `pos_label` https://github.com/scikit-learn/scikit-learn/pull/16061#discussion_r701993621 which can be done a follow up PR.

thomasjpfan committed:
TST Fixes test error

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
FIX Uses labelencoder

glemaitre commented on doc/whats_new/v1.1.rst:
```diff
@@ -239,6 +239,13 @@ Changelog
   get accurate results when the number of features is large.
   :pr:`21109` by :user:`Smile <x-shadow-man>`.
 
+- |Fix| :class:`decomposition.FastICA` now validates input parameters in `fit` instead of `__init__`.
```
I think that these entries should be removed.

glemaitre reviewed (approved):
LGTM The PR still looks good.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into plot_decision_bound

thomasjpfan committed:
REV Remove unneeded whats new item

jeremiedbb committed:
Update sklearn/inspection/_plot/decision_boundary.py

jeremiedbb committed:
Update sklearn/inspection/_plot/decision_boundary.py

jeremiedbb committed:
Merge branch 'main' into plot_decision_bound

jeremiedbb reviewed (approved):
LGTM.

The resolution of the grid seems a little bit lower than the current one in several examples, which makes the boundary a little bit more dented but it's a really subtle difference.

I agree with what was said previously, adding the possibility to pass the pos_label can be done in a separate PR.

Does anyone have second thoughts before we merge ?
